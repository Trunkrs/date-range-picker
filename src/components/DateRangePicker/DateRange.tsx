import { createContext } from 'react'

export interface DateRangeContext {
  dateStart?: Date | null
  dateEnd?: Date | null
}

export const DateRange = createContext<DateRangeContext>({
  dateStart: null,
  dateEnd: null,
})

export default DateRange
