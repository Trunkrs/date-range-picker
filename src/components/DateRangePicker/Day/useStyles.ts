import createStyles from '@material-ui/core/styles/createStyles'
import makeStyles from '@material-ui/core/styles/makeStyles'
import { Theme } from '@material-ui/core/styles'

const styles = ({ palette }: Theme) =>
  createStyles({
    day: {
      position: 'relative',
      textAlign: 'center',
      width: '100%',
      height: '100%',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
      overflow: 'hidden',
      borderTop: '1px solid white',
      borderBottom: '1px solid white',
    },
    hovering: {
      '&:hover, &:focus': {
        border: `4px solid white`,
        backgroundColor: palette.trunkrs.grayscale.purgatory,
        borderRadius: '50%',
        width: '100%',
        height: '100%',
        justifyContent: 'center',
        display: 'flex',
        alignItems: 'center',
      },
    },
    outsideDay: {
      cursor: 'default',
    },
    selected: {
      backgroundColor: palette.trunkrs.secondary.lightViolet.fade,
    },
    start: {
      borderTopLeftRadius: '50%',
      borderBottomLeftRadius: '50%',
    },
    end: {
      borderTopRightRadius: '50%',
      borderBottomRightRadius: '50%',
    },
    selectedDay: {
      width: '100%',
      height: '100%',
      backgroundColor: palette.trunkrs.secondary.violet.base,
      alignItems: 'center',
      display: 'flex',
      justifyContent: 'center',
      borderRadius: '50%',
      color: palette.common.white,
      position: 'relative',
      border: `4px solid ${palette.trunkrs.secondary.lightViolet.fade}`,
    },
    hidden: {
      opacity: 0,
    },
  })

export default makeStyles(styles)
