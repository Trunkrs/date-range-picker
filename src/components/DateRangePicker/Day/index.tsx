import React, { useCallback } from 'react'

import clsx from 'clsx'

import ButtonBase from '@material-ui/core/ButtonBase'
import Typography from '../../Typography'

import useStyles from './useStyles'
import useDateHighlight from '../../../hooks/useDateHighlight'

export interface DayProps {
  children: number
  date: Date
  show?: boolean
  calendarMonth?: Date
  onClick?: (value: Date) => void
  outsideDayType?: 'prev' | 'next'
}

export const Day: React.FC<DayProps> = ({
  outsideDayType = false,
  calendarMonth,
  show = true,
  children,
  onClick,
  date,
}) => {
  const { isHighlighted, isStart, isEnd } = useDateHighlight(
    date,
    outsideDayType,
    calendarMonth,
  )
  const classes = useStyles()

  const handleClick = useCallback(() => {
    if (onClick) {
      onClick(date)
    }
  }, [onClick, date])

  const theDate = React.useMemo(
    () => (
      <Typography
        className={clsx(
          (isStart || isEnd) && !outsideDayType && classes.selectedDay,
        )}
        variant="text"
        color={outsideDayType ? 'grey' : 'indigo'}
      >
        {children}
      </Typography>
    ),
    [classes, children, isStart, isEnd, outsideDayType],
  )

  return (
    <ButtonBase
      disabled={!!outsideDayType}
      className={clsx(
        classes.day,
        !isHighlighted &&
          !outsideDayType &&
          !isStart &&
          !isEnd &&
          classes.hovering,
        outsideDayType && classes.outsideDay,
        isHighlighted && classes.selected,
        isStart && classes.start,
        isEnd && classes.end,
        !show && classes.hidden,
      )}
      onClick={handleClick}
    >
      {theDate}
    </ButtonBase>
  )
}

export default React.memo(Day)
