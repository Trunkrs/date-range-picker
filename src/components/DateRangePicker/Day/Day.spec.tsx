import React from 'react'

import { createShallow, createMount } from '@material-ui/core/test-utils'
import ButtonBase from '@material-ui/core/ButtonBase'

import Day, { DayProps } from '.'
import DateRange from '../DateRange'

const createProps = (merge: Partial<DayProps> = {}): DayProps =>
  ({
    children: new Date(2020, 9, 20, 16, 0, 0).getDate(),
    date: new Date(2020, 9, 20, 16, 0, 0),
    onClick: jest.fn(),
    ...merge,
  } as DayProps)

describe('<Day />', () => {
  const shallow = createShallow()
  const mount = createMount()

  it('should match snapshot', () => {
    const wrapper = shallow(<Day {...createProps()} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should call the onClick fn when day is clicked', () => {
    const props = createProps()
    const wrapper = mount(
      <DateRange.Provider
        value={{
          dateStart: new Date(2020, 9, 20, 0, 0, 0),
          dateEnd: new Date(2020, 9, 23, 0, 0, 0),
        }}
      >
        <Day {...props} />
      </DateRange.Provider>,
    )
    wrapper.find(ButtonBase).simulate('click')
    expect(props.onClick).toHaveBeenCalled()
  })
})
