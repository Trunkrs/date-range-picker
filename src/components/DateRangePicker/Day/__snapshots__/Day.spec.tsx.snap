// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Day /> should match snapshot 1`] = `
<WithStyles(ForwardRef(ButtonBase))
  className="makeStyles-day-1 makeStyles-hovering-2"
  disabled={false}
  onClick={[Function]}
>
  <WithStyles(Typography)
    className=""
    color="indigo"
    variant="text"
  >
    20
  </WithStyles(Typography)>
</WithStyles(ForwardRef(ButtonBase))>
`;
